Jobs.Fetcher.TikTok/Helpers/DbWriter.cs

```diff
@@ -17,82 +17,104 @@ namespace Jobs.Fetcher.TikTok.Helpers {
+                    var oldEntry = dbContext.Challenges.Find(newEntry.Id);
+                    Upsert<Challenge, DataLakeTikTokContext>(oldEntry, newEntry, dbContext, logger);
+                }
+                dbContext.SaveChanges();
             }
-            dbContext.SaveChanges();
         }
 
         public static void WriteEffectStickers(List<EffectSticker> newEntries, DataLakeTikTokContext dbContext, Logger logger) {
-            foreach (var newEntry in newEntries) {
-                var oldEntry = dbContext.EffectStickers.Find(newEntry.Id);
-                Upsert<EffectSticker, DataLakeTikTokContext>(oldEntry, newEntry, dbContext, logger);
+            lock (DbLock) {
+                foreach (var newEntry in newEntries) {
+                    var oldEntry = dbContext.EffectStickers.Find(newEntry.Id);
+                    Upsert<EffectSticker, DataLakeTikTokContext>(oldEntry, newEntry, dbContext, logger);
+                }
+                dbContext.SaveChanges();
             }
-            dbContext.SaveChanges();
         }
 
         public static void WriteMusic(Music newEntry, DataLakeTikTokContext dbContext, Logger logger) {
-            var oldEntry = dbContext.Music.Find(newEntry.Id);
-            Upsert<Music, DataLakeTikTokContext>(oldEntry, newEntry, dbContext, logger);
-            dbContext.SaveChanges();
+            lock (DbLock) {
+                var oldEntry = dbContext.Music.Find(newEntry.Id);
+                Upsert<Music, DataLakeTikTokContext>(oldEntry, newEntry, dbContext, logger);
+                dbContext.SaveChanges();
+            }
         }
 
         public static void WritePost(Post newEntry, DataLakeTikTokContext dbContext, Logger logger) {
             var now = DateTime.UtcNow;
-            var oldEntry = dbContext.Posts.SingleOrDefault(m => m.Id == newEntry.Id && m.ValidityStart <= now && m.ValidityEnd > now);
-            Insert<Post, DataLakeTikTokContext>(oldEntry, newEntry, dbContext, logger);
-            dbContext.SaveChanges();
+            lock (DbLock) {
+                var oldEntry = dbContext.Posts.SingleOrDefault(m => m.Id == newEntry.Id && m.ValidityStart <= now && m.ValidityEnd > now);
+                Insert<Post, DataLakeTikTokContext>(oldEntry, newEntry, dbContext, logger);
+                dbContext.SaveChanges();
+            }
         }
 
         public static void WritePostStats(PostStats newEntry, DataLakeTikTokContext dbContext, Logger logger) {
             var now = DateTime.UtcNow;
-            var oldEntry = dbContext.Stats.SingleOrDefault(m => m.PostId == newEntry.PostId && m.ValidityStart <= now && m.ValidityEnd > now);
-            Insert<PostStats, DataLakeTikTokContext>(oldEntry, newEntry, dbContext, logger);
-            dbContext.SaveChanges();
+            lock (DbLock) {
+                var oldEntry = dbContext.Stats.SingleOrDefault(m => m.PostId == newEntry.PostId && m.ValidityStart <= now && m.ValidityEnd > now);
```

Comment:
I just realized a big mistake that can be breaking everything. We are updating every last values and not just on the update date.
```suggestion
                var oldEntry = dbContext.Stats.SingleOrDefault(m => m.PostId == newEntry.PostId && m.ValidityStart <= now && m.ValidityEnd > now && m.EventDate == newEntry.EventDate);
```

Reply:
Woah, you're right on that! Good catch! I think it wouldn't update every last value, because of SingleOrDefault. But, in fact, the row to be updated would be a random one, from all those returned by the conditional